### ISSUE TITLE
YOLOv8 Inference has a very serious bug!!!

### ISSUE THREAD

**Issue by Leeinsn** [ultralytics/ultralytics#2473]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I trained a model with Yolov8 and found that it would **crash my computer** when inferring. 

After a few hours of debugging, I found that it was due to **memory not being freed in time** when inferring. Specifically, since the number of images I need to inference is relatively large, my memory can't handle such a large overhead. Therefore, when inferring to more than a thousand images, the nms of one image even took an hour. And because the memory was full, the running of other programs of the computer was disturbed. 

I think it's a pretty big bug, and I wish it could be brought to attention and fixed as soon as possible.

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by Leeinsn** [comment]
ok, my inference code is as following

```python
import cv2
from ultralytics import YOLO

model_path = '/home/zhangyt/runs/detect/train2/weights/best.pt'
test_path = '/home/zhangyt/giga/PANDA-CUT/testA/test_18'

model = YOLO(model_path)
model.predict(
    source=test_path,
    save=True,
    imgsz=1280,
    # conf=0.25,  # no nms
    # iou=0.7,
    show=False,
    save_txt=True,
    save_conf=True,
    max_det=300,
    augment=False
    )
```
Maybe something wrong with my code? Thanks a lot!!!

**Comment by glenn-jocher** [comment]
@Leeinsn, thanks for providing the code you are using. It seems that you are correctly using the `predict()` method from YOLOv8. Because you are using `save=True` and `show=False`, the images generated during the prediction are being saved to disk but not shown on the screen. 

The likely cause of your issue is that your computer has insufficient memory to handle the number of images being processed in inference. When dealing with a large number of images it's advisable to use a more powerful machine with enough RAM, if possible. 

One possible solution is to batch process your images, which will help limit the memory overhead of the inference task. Alternatively, you could try reducing the image size by passing a smaller value for `imgsz` argument to the `predict()` function.

**Comment by Leeinsn** [comment]
It turned out to be my fault for setting the `save=True`. Thank you for your patient answer, it worked well for me.

**Comment by glenn-jocher** [comment]
@Leeinsn great! Also see streaming mode here for processing single images at a time rather than a group:
https://docs.ultralytics.com/modes/predict/

**Comment by AyushExel** [comment]
@Leeinsn you could also just use `stream=True` to avoid memory leakage issues

**Comment by shv07** [comment]
I am facing the same issue with pretrained yolov8x model. I am using the predict method with default arguments:-
```python
results = model.predict(img, conf=conf, device='cpu') 
```
Library versions: latest available for Ubuntu

The RAM usage keeps increasing marginally with more and more inferences. I tried by exporting the model to onnx  and then using it. The issue remains the same.

**Comment by glenn-jocher** [comment]
Hi @<user>, thank you for bringing this to our attention. The issue you are facing might be due to a memory leak in the `predict()` method when processing many images using default arguments. To mitigate this issue, we would suggest trying the `stream=True` option for the `predict()` method. 

Setting `stream=True` will process a single image at a time before moving onto the next, which should reduce the amount of memory consumed during inference for a set of images.

If this doesn't work, alternatives are resizing images to reduce their resolution before processing, or processing images in batches to reduce the overall memory footprint.

Please let us know if this helps resolve the issue or if you have any other questions we could assist you with.

**Comment by shv07** [comment]
Thanks for your response @glenn-jocher. Using ```stream=True``` does not help with the issue (I process only one image at a time though). Even forcefully deleting the ```results``` object does not help. As for image resolution, I am already passing all images at a 320x240. My primary concern here is continuous increases in RAM with each inference.

**Comment by glenn-jocher** [comment]
Hi @<username>, I'm sorry to hear that using `stream=True` did not help with the RAM usage issue you are facing when using YOLOv8. In this case, one possible explanation for an increase in RAM usage could be due to the autograd engine in PyTorch accumulating gradients of all operations to enable computation of gradients at the end of a set of operations. To access intermediate results while training and inference in PyTorch without accumulating these gradients, you can use the `torch.no_grad()` context manager, which disables gradient computation.

Therefore, you can try using the `torch.no_grad()` context manager to disable autograd computations on your model predictions. Another approach you can try is to run inference on a CPU in case it is being run on GPU.

Please let me know if you have any other questions or clarifications.

**Comment by Laitaps** [comment]
I can confirm this same issue.  I get very short duration spikes in memory consumption after each predict call.  They get larger with each pass until they either consume all available memory (16GB) or eventually settle down.  Image size is 1024.  I have tried setting stream=True, turning off saving of images and @torch.no_grad().  No solution seems to work.  Conducting inference on my CPU.  I will test on my Orin Nano to see if that environment behaves differently.

**Comment by Laitaps** [comment]
def sv_detect(source:List[str], weights:str, device:str='cpu') -> List:
    print("CNN object detection on the following:")
    for i, url in enumerate(source):
        print(f"{i}: {url}")
    augment = True
    conf = 0.70
    iou = 0.75
    classes = None
    half = False
    model = YOLO(weights)
    imgsz = (1024, 1024)

    print("Loading images...")
    lst_as_img = []
    for _, url in enumerate(source):
        lst_as_img.append(url_to_image(url))

    print("predicting")
    results = model.predict(
        source=lst_as_img,
        device=device,
        imgsz=imgsz,
        save=True,
        save_txt=True,
        augment=augment,
        conf=conf,
        iou=iou,
        classes=classes,
        half=half)

**Comment by Laitaps** [comment]
``def sv_detect(source:List[str], weights:str, device:str='cpu') -> List:
    print("CNN object detection on the following:")
    for i, url in enumerate(source):
        print(f"{i}: {url}")
    augment = True
    conf = 0.70
    iou = 0.75
    classes = None
    half = False
    model = YOLO(weights)
    imgsz = (1024, 1024)

    print("Loading images...")
    lst_as_img = []
    for _, url in enumerate(source):
        lst_as_img.append(url_to_image(url))

    print("predicting")
    results = model.predict(
        source=lst_as_img,
        device=device,
        imgsz=imgsz,
        save=True,
        save_txt=True,
        augment=augment,
        conf=conf,
        iou=iou,
        classes=classes,
        half=half)

**Comment by Laitaps** [comment]
```
def sv_detect(source:List[str], weights:str, device:str='cpu') -> List:
    print("CNN object detection on the following:")
    for i, url in enumerate(source):
        print(f"{i}: {url}")
    augment = True
    conf = 0.70
    iou = 0.75
    classes = None
    half = False
    model = YOLO(weights)
    imgsz = (1024, 1024)

    print("Loading images...")
    lst_as_img = []
    for _, url in enumerate(source):
        lst_as_img.append(url_to_image(url))

    print("predicting")
    results = model.predict(
        source=lst_as_img,
        device=device,
        imgsz=imgsz,
        save=True,
        save_txt=True,
        augment=augment,
        conf=conf,
        iou=iou,
        classes=classes,
        half=half)
```

**Comment by Laitaps** [comment]
Turning off augmentation does not resolve the issue either.

**Comment by pderrenger** [comment]
@Laitaps thanks for the additional context. The code you've provided seems to align with the recommendations we've discussed earlier, such as passing `stream=True` and utilizing `torch.no_grad()`. 

Given the persistent memory spikes despite these adjustments, I recommend profiling the code to pinpoint the specific operations or components responsible for the increasing RAM usage. This can be done using profiling tools in Python, or by systematically commenting out portions of your code to detect the precise source of the memory issue.

Additionally, testing the code on the Orin Nano to observe its behavior in a different environment is an excellent idea, as it can help identify potential environment-specific factors contributing to the issue.

Please let me know if you need further assistance or if you have any updates after testing on the Orin Nano.

**Comment by Laitaps** [comment]
> @Laitaps thanks for the additional context. The code you've provided seems to align with the recommendations we've discussed earlier, such as passing `stream=True` and utilizing `torch.no_grad()`. 
> 
> Given the persistent memory spikes despite these adjustments, I recommend profiling the code to pinpoint the specific operations or components responsible for the increasing RAM usage. This can be done using profiling tools in Python, or by systematically commenting out portions of your code to detect the precise source of the memory issue.
> 
> Additionally, testing the code on the Orin Nano to observe its behavior in a different environment is an excellent idea, as it can help identify potential environment-specific factors contributing to the issue.
> 
> Please let me know if you need further assistance or if you have any updates after testing on the Orin Nano.

Thanks for your reply.  I got temporarily pulled to another ML project.  When I return to this one in a few weeks, I should have more information.

**Comment by Laitaps** [comment]
Thanks for your reply.  I should have more information for you when I return to this project in a couple weeks.

**Comment by glenn-jocher** [comment]
@Laitaps, you're welcome! I appreciate your proactive approach to resolving the issue, and I look forward to hearing more about your findings in the coming weeks.

If you run into any other challenges or have additional questions when you return to this project, please don't hesitate to reach out. Meanwhile, if there's anything specific that we can assist with, let us know.

Good luck with your testing, and we hope the situation improves on the Orin Nano. Take care and we'll be here when you're ready to continue the discussion. 🛠️🔍

**Comment by Laitaps** [comment]
I have rotated back to this project and can confirm the same issue on the Orin Nano with JetPack 6 installed.  I will dig deeper and provide additional details.

**Comment by glenn-jocher** [comment]
@Laitaps thank you for the update. It's helpful to know that the issue persists across different platforms. We'll await your detailed findings to further investigate the matter. If you discover any specific patterns or triggers for the memory spikes, that information will be crucial for diagnosing the problem. 🕵️‍♂️🔍
